### PR TITLE
Add CMake Windows support

### PR DESCRIPTION
**Pull request by dnakamura**
- Create new fixpath2.sh tool, which can handle /cygrive/ in the middle
  of a path

- Export MSVC variables when calling cmake

Signed-off-by: Devin Nakamura <devinn@ca.ibm.com>

### CODE REVIEW

**Comment by dnakamura** [comment]
Marked as draft, as my test build is still running.
Note a couple of patches are required for the windows build to work:
eclipse/openj9#10467
eclipse/openj9#10466
eclipse/omr#5505

**Comment by dnakamura** [comment]
note: there are also some remaining issues with DDR (which need to be fixed on the omr side)
@keithc-ca  can you take a look at this?

File: closed/OpenJ9.gmk [hunk 1]
```diff
@@ -379,14 +384,15 @@ ifeq (true,$(OPENJ9_ENABLE_CMAKE))
...
+    CMAKE_ARGS += -DJ9VM_OPT_CUDA=ON -DOMR_CUDA_HOME="$(call UnixPath,$(CUDA_HOME))"
   else # OPENJ9_ENABLE_CUDA
     CMAKE_ARGS += -DJ9VM_OPT_CUDA=OFF
   endif # OPENJ9_ENABLE_CUDA
 
 $(OUTPUTDIR)/vm/cmake.stamp :
 	@$(MKDIR) -p $(@D)
-	cd $(@D) && $(CMAKE) $(CMAKE_ARGS) $(OPENJ9_TOPDIR)
+	cd $(@D) && export $(EXPORT_MSVS_ENV_VARS) > /dev/null \
+		&& $(CMAKE) $(CMAKE_ARGS) $(OPENJ9_TOPDIR)
```
Comment (keithc-ca) [inline closed/OpenJ9.gmk:395]: This should not use `export`, eliminating the need to ignore errors; it should be:
```
cd $(@D) && $(EXPORT_MSVS_ENV_VARS) $(CMAKE) $(CMAKE_ARGS) $(OPENJ9_TOPDIR)
```

File: closed/autoconf/custom-hook.m4 [hunk 2]
```diff
@@ -667,17 +667,20 @@ AC_DEFUN([OPENJ9_GENERATE_TOOL_WRAPPERS],
...
+  OPENJ9_GENERATE_TOOL_WRAPPER([cl], [$FIXPATH2], [$CC])
+  OPENJ9_GENERATE_TOOL_WRAPPER([lib], [$FIXPATH2], [$AR])
+  OPENJ9_GENERATE_TOOL_WRAPPER([link], [$FIXPATH2], [$LD])
+  OPENJ9_GENERATE_TOOL_WRAPPER([ml], [$FIXPATH2], [$MSVC_BIN_DIR/ml])
+  OPENJ9_GENERATE_TOOL_WRAPPER([ml64], [$FIXPATH2], [$MSVC_BIN_DIR/ml64])
+  OPENJ9_GENERATE_TOOL_WRAPPER([rc], [$FIXPATH2], [$RC])
+  OPENJ9_GENERATE_TOOL_WRAPPER([mc], [$FIXPATH2], [$SDK_BIN_DIR/mc])
+  OPENJ9_GENERATE_TOOL_WRAPPER([nasm], [$FIXPATH2], [$NASM])
+
+  # fixpath2 cant handle classpaths, or @<file_name> arguments
```
Comment (keithc-ca) [inline closed/autoconf/custom-hook.m4:682]: typo: cant -> can't

File: closed/fixpath2.sh [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+cmd="$1" # Our first arg is the name of the command to be invoked
+shift
+
+new_args=()
+for arg in "$@"; do
+    new_args+=( "$(fix_argument "$arg")" )
+done
+
+"$cmd" "${new_args[@]}"
```
Comment (keithc-ca) [inline closed/fixpath2.sh:55]: This should not use `echo` nor `printf`:
```
cmd=()
for arg in "$@"; do
  case "$arg" in
    -I/*)
      arg="-I$(cygpath -w "${arg:2}")"
      ;;
    /DEF:/*)
      arg="/DEF:$(cygpath -w "${arg:5}")"
      ;;
    /pdb:/*)
      arg="/pdb:$(cygpath -w "${arg:5}")"
      ;;
    /cygdrive/*)
      arg="$(cygpath -w "$arg")"
      ;;
    *)
      ;;
  esac
  cmd+=("$arg")
done

"${cmd[@]}"
```

**Comment by keithc-ca** [comment]
@dnakamura How did you test this? When I tried with:
```
bash configure \
    --disable-ccache \
    --disable-ddr \
    --with-boot-jdk=/cygdrive/c/space/bootjdk11 \
    --with-cmake \
    --with-toolchain-version=2017
```
It failed:
```
-- The C compiler identification is unknown
-- The CXX compiler identification is unknown
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - failed
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - failed
-- Using BOOT_JDK: '/cygdrive/c/space/bootjdk11'
CMake Warning at /cygdrive/c/space/jdk11/omr/cmake/modules/OmrDetectSystemInformation.cmake:186 (message):
  OMR: C Compiler ID is not set.  Did you call omr_detect_system_information
  before C was enabled?
Call Stack (most recent call first):
  /cygdrive/c/space/jdk11/omr/cmake/modules/OmrPlatform.cmake:32 (omr_detect_system_information)
  runtime/cmake/platform.cmake:24 (include)
  runtime/CMakeLists.txt:48 (include)


-- OMR: The CPU architecture is x86
-- OMR: The OS is win
-- OMR: The tool configuration is
-- OMR: The target data size is 64
-- Looking for dladdr
-- Looking for dladdr - not found
-- The ASM compiler identification is unknown
-- Found assembler: /cygdrive/c/space/jdk11/build/windows-x86_64-normal-server-release-cmake/tools/cl
-- Warning: Did not find file Compiler/-ASM
CMake Error at /cygdrive/c/space/jdk11/omr/cmake/modules/OmrPlatform.cmake:45 (include):
  include could not find load file:

    platform/toolcfg/
Call Stack (most recent call first):
  runtime/cmake/platform.cmake:24 (include)
  runtime/CMakeLists.txt:48 (include)


CMake Warning at /cygdrive/c/space/jdk11/omr/cmake/modules/OmrAssert.cmake:58 (message):
  Assertion failed: DEFINED OMR_WARNING_AS_ERROR_FLAG
```

File: closed/OpenJ9.gmk (hunk 1 above)
Comment (dnakamura) [reply closed/OpenJ9.gmk:395]: That is what I initially tried but I was getting the same errors about unidentified compilers. Although now I am suspecting its an unrelated intermittent issue

**Comment by dnakamura** [comment]
I was having the same issue, but I thought I had resolved it. As a work around if you launch cygwin from the visual studio console  it should fix the issue (note you will need to delete the cache as cmake wont try to re-identify the compiler)

**Comment by dnakamura** [comment]
When applying fixes I also noticed a bug in the fixpath implementation you provided, in that it also converts the path of the compiler to a windows style path. I'm not sure if you made that change locally before you tried to build, but it would also cause the same failure to identify the compiler. If that's not it looking at vm/CMakeFiles/CMakeError.log may provide some insight as to the issue

**Comment by dnakamura** [comment]
Added tool wrapper for clang to support eclipse/openj9#10499

File: closed/autoconf/custom-hook.m4 [hunk 4]
```diff
@@ -649,16 +652,16 @@ AC_DEFUN([CONFIGURE_OPENSSL],
 ])
 
 # Create a tool wrapper for use by cmake.
-# Consists of a shell script which wraps commands with an invocation of fixpath.
-# OPENJ9_GENERATE_TOOL_WRAPER(<name_of_wrapper>, <command_to_call>)
+# Consists of a shell script which wraps commands with an invocation a wrapper command
```
Comment (keithc-ca) [inline closed/autoconf/custom-hook.m4:655]: Something seems missing here, perhaps you meant to say 'invocation _of_ a wrapper command' (with a period).

File: closed/OpenJ9.gmk [hunk 5]
```diff
@@ -70,6 +70,10 @@ endif
 ifeq (windows,$(OPENJDK_TARGET_OS))
   # convert unix path to windows path
   FixPath = $(shell $(CYGPATH) -m $1)
+
+  # convert windows path to unix path
+  UnixPath = $(shell $(CYGPATH) -u $1)
+
```
Comment (keithc-ca) [inline closed/OpenJ9.gmk:76]: Please omit the blank lines (73 & 76).

File: closed/autoconf/custom-hook.m4 [hunk 6]
```diff
@@ -667,17 +670,21 @@ AC_DEFUN([OPENJ9_GENERATE_TOOL_WRAPPERS],
...
+  OPENJ9_GENERATE_TOOL_WRAPPER([ml64], [$FIXPATH2], [$MSVC_BIN_DIR/ml64])
+  OPENJ9_GENERATE_TOOL_WRAPPER([rc], [$FIXPATH2], [$RC])
+  OPENJ9_GENERATE_TOOL_WRAPPER([mc], [$FIXPATH2], [$SDK_BIN_DIR/mc])
+  OPENJ9_GENERATE_TOOL_WRAPPER([nasm], [$FIXPATH2], [$NASM])
+  OPENJ9_GENERATE_TOOL_WRAPPER([clang], [$FIXPATH2], [$OPENJ9_CLANG])
+
+  # fixpath2 can't handle classpaths, or @<file_name> arguments
+  OPENJ9_GENERATE_TOOL_WRAPPER([java], [$FIXPATH], [$JAVA])
+  OPENJ9_GENERATE_TOOL_WRAPPER([jar], [$FIXPATH], [$JAR])
+  OPENJ9_GENERATE_TOOL_WRAPPER([javac], [$FIXPATH],[$JAVAC])
```
Comment (keithc-ca) [inline closed/autoconf/custom-hook.m4:689]: Since you're changing all these lines, could you please sort them in each group by command name.

File: closed/fixpath2.sh [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+#
+# This code is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+# version 2 for more details (a copy is included in the LICENSE file that
+# accompanied this code).
+#
+# You should have received a copy of the GNU General Public License version
+# 2 along with this work; if not, see <http://www.gnu.org/licenses/>.
+#
```
Comment (keithc-ca) [inline closed/fixpath2.sh:22]: Please remove lines 6 and 22; extra blank lines in the copyright notice are not needed.

File: closed/autoconf/toolchain-win.cmake.in [hunk 8]
```diff
@@ -26,6 +26,7 @@ set(CMAKE_SYSTEM_PROCESSOR "@OPENJDK_TARGET_CPU@")
 set(tool_dir "@OPENJ9_TOOL_DIR@")
 set(CMAKE_C_COMPILER "${tool_dir}/cl" CACHE FILEPATH "")
 set(CMAKE_CXX_COMPILER "${tool_dir}/cl" CACHE FILEPATH "")
+set(CMAKE_J9VM_C_ALT_COMPILER "${tool_dir}/clang" CACHE FILEPATH "")
```
Comment (keithc-ca) [inline closed/autoconf/toolchain-win.cmake.in:29]: This should be deferred to a separate pull request, or be updated once a final language name is chosen (see https://github.com/eclipse/openj9/pull/10499#discussion_r483142852).

**Comment by keithc-ca** [comment]
Jenkins test sanity win,wincm jdk11

**Comment by keithc-ca** [comment]
There seem to be problems with DDR. Lots of errors like this:
```
14:49:09  omrddr/__/build/windows-x86_64-normal-server-release/vm/runtime/omr/omrcfg.stub.h(2): fatal error C1083: Cannot open include file: '/cygdrive/c/Users/jenkins/workspace/Build_JDK11_x86-64_windows_cm_Personal/build/windows-x86_64-normal-server-release/vm/runtime/omr/omrcfg.h': No such file or directory
```
The compiler doesn't understand cygwin paths.

**Comment by dnakamura** [comment]
see eclipse/omr#4666

**Comment by dnakamura** [comment]
Also, eclipse/openj9#10505 is needed as ddr errors out trying to handle some internal debug types from the msvc standard library

**Comment by keithc-ca** [comment]
Marked as draft; mark as ready for review when you have tested with all necessary changes (either merged or documented here).

**Comment by dnakamura** [comment]
@keithc-ca works with  eclipse/openj9#10499

**Comment by keithc-ca** [comment]
Jenkins test sanity x86-64_windows_cm,x86-64_windows_uma jdk11 depends eclipse/openj9#10499

**Comment by keithc-ca** [comment]
Jenkins test sanity x86-64_windows,x86-64_windows_cm jdk11 depends eclipse/openj9#10499

**Comment by keithc-ca** [comment]
Building jars for testing the cmake build fails:
```
15:36:02        [jar] Building jar: F:\Users\****\workspace\Test_openjdk11_j9_sanity.functional_x86-64_windows_cm_Personal_testList_0\jvmtest\functional\JLM_Tests\jlm_tests.jar
15:36:02  Unhandled exception
15:36:02  Type=Segmentation error vmState=0x00000000
15:36:02  Windows_ExceptionCode=c0000005 J9Generic_Signal=00000004 ExceptionAddress=00007FFDD13B26C3 ContextFlags=0010005f
15:36:02  Handler1=00007FFDD2265D30 Handler2=00007FFDD3F7AA80 InaccessibleReadAddress=00000000000001FF
15:36:02  RDI=00000000FFD602A7 RSI=00000000FFD602A8 RAX=00000000000001FF RBX=0000000000000000
15:36:02  RCX=00000000FFD602A7 RDX=00000000FFD60548 R8=0000000000000004 R9=000000000000005C
15:36:02  R10=0000000000000014 R11=00000000FFD60548 R12=0000000000000014 R13=00000000FFD60548
15:36:02  R14=0000000000000003 R15=0000000000000000
15:36:02  RIP=00007FFDD13B26C3 RSP=000000000010DA60 RBP=0000000000019A00 GS=002B
15:36:02  FS=0053 ES=002B DS=002B
15:36:02  XMM0 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:02  XMM1 003a003a003a003a (f: 3801146.000000, d: 1.446347e-307)
15:36:02  XMM2 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:02  XMM3 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:02  XMM4 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:02  XMM5 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:02  XMM6 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:02  XMM7 0061006500720063 (f: 7471203.000000, d: 7.565937e-307)
15:36:02  XMM8 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:02  XMM9 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:02  XMM10 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:02  XMM11 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:02  XMM12 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:02  XMM13 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:02  XMM14 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:02  XMM15 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:02  Module=F:\Users\****\workspace\Test_openjdk11_j9_sanity.functional_x86-64_windows_cm_Personal_testList_0\openjdkbinary\j2sdk-image\bin\compressedrefs\j9jit29.dll
15:36:02  Module_base_address=00007FFDD0DC0000 Offset_in_DLL=00000000005f26c3
15:36:02  Target=2_90_20200929_2 (Windows Server 2012 R2 6.3 build 9600)
15:36:02  CPU=amd64 (8 logical CPUs) (0x1ffb9c000 RAM)
15:36:02  ----------- Stack Backtrace -----------
15:36:02  Java_java_lang_invoke_MutableCallSite_invalidate+0x524ae3 (0x00007FFDD13B26C3 [j9jit29+0x5f26c3])
15:36:02  Java_java_lang_invoke_MutableCallSite_invalidate+0x50627b (0x00007FFDD1393E5B [j9jit29+0x5d3e5b])
15:36:02  (0x000000000000005C)
15:36:02  ---------------------------------------
```
and
```
15:36:13        [jar] Building jar: F:\Users\****\workspace\Test_openjdk11_j9_sanity.functional_x86-64_windows_cm_Personal_testList_1\jvmtest\functional\JIT_Test\jitt.jar
15:36:13  Unhandled exception
15:36:13  Type=Segmentation error vmState=0x00000000
15:36:13  Windows_ExceptionCode=c0000005 J9Generic_Signal=00000004 ExceptionAddress=00007FF87C701930 ContextFlags=0010005f
15:36:13  Handler1=00007FF87DC55D30 Handler2=00007FF87F96AA80 InaccessibleReadAddress=0000000000000014
15:36:13  RDI=0000000000000000 RSI=0000000000000014 RAX=0000000000019A04 RBX=00000000FFD3E310
15:36:13  RCX=0000000000019A00 RDX=00000000FFD3E450 R8=0000000000000004 R9=000000000000005C
15:36:13  R10=0000000000000004 R11=00000000FFD3E1C0 R12=00000000FFD3E450 R13=0000000000019A04
15:36:13  R14=0000000000000055 R15=000000000000005C
15:36:13  RIP=00007FF87C701930 RSP=000000000010E008 RBP=0000000000019A00 GS=002B
15:36:13  FS=0053 ES=002B DS=002B
15:36:13  XMM0 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:13  XMM1 003a003a003a003a (f: 3801146.000000, d: 1.446347e-307)
15:36:13  XMM2 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:13  XMM3 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:13  XMM4 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:13  XMM5 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:13  XMM6 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:13  XMM7 0061006500720063 (f: 7471203.000000, d: 7.565937e-307)
15:36:13  XMM8 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:13  XMM9 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:13  XMM10 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:13  XMM11 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:13  XMM12 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:13  XMM13 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:13  XMM14 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:13  XMM15 0000000000000000 (f: 0.000000, d: 0.000000e+00)
15:36:13  Module=F:\Users\****\workspace\Test_openjdk11_j9_sanity.functional_x86-64_windows_cm_Personal_testList_1\openjdkbinary\j2sdk-image\bin\compressedrefs\j9jit29.dll
15:36:13  Module_base_address=00007FF87C700000 Offset_in_DLL=0000000000001930
15:36:13  Target=2_90_20200929_2 (Windows Server 2012 R2 6.3 build 9600)
15:36:13  CPU=amd64 (8 logical CPUs) (0x1ffb9c000 RAM)
15:36:13  ----------- Stack Backtrace -----------
15:36:13  (0x00007FF87C701930 [j9jit29+0x1930])
15:36:13  (0x00007FF868E32C01)
15:36:13  (0x00000000FFD3E450)
15:36:13  ---------------------------------------
```

**Review by keithc-ca — APPROVED** [review]

**Comment by keithc-ca** [comment]
Merging despite test failures - see https://github.com/eclipse/openj9/pull/10499#issuecomment-700983927.
